Diagnose unannotated C++ APIs returning SWIFT_SHARED_REFERENCE at Swift call sites

fahadnayyar opened the pull request:
This patch improves the warning for C++ APIs returning `SWIFT_SHARED_REFERENCE` types but not annotated with either `SWIFT_RETURNS_RETAINED` or `SWIFT_RETURNS_UNRETAINED` in the following ways:

1. The warning for missing `SWIFT_RETURNS_(UN)RETAINED` annotations is now emitted on Swift use sites, rather than while importing the API (func/method decls).
- This logic is now implemented as a Misl Diagnostic in function `diagnoseCxxFunctionCalls` in file lib/Sema/MiscDiagnostics.cpp. 
- The warning is now triggered only when the API is actually used, which reduces noise in large C++ headers.

rdar://150800115

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

Xazax-hun commented:
> The warning is now triggered only when the API is actually used, which reduces noise in large C++ headers.

Do you have some examples where we ended up warning for functions that are not used? We generally try to be samewhat lazy with importing functions (with some exceptions like operators, construtors), and only import something when it is actually used from Swift. I am wondering, if the extra warnings are telltale that we might want to make something more lazy in the importer logic.

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3637,10 +3636,15 @@ namespace {
-                Diagnostic(diag::no_returns_retained_returns_unretained, decl),
-                decl->getLocation());
+            ASTContext &ctx = Impl.SwiftContext;
+            std::string msg =
```
```suggestion
            String Rerf msg =
```

Xazax-hun replied:
You don't need to heap allocate the message.

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3637,10 +3636,15 @@ namespace {
+                "It should be annotated with either SWIFT_RETURNS_RETAINED or "
+                "SWIFT_RETURNS_UNRETAINED as it is returning a "
+                "SWIFT_SHARED_REFERENCE";
+            auto message = ctx.AllocateCopy(StringRef(msg));
```
I think you can also get rid of `ctx.AllocateCopy` since we are dealing with a static string here.

fahadnayyar commented:
@swift-ci please smoke test

j-hui commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -5001,6 +5001,7 @@ bool ClangImporter::Implementation::emitDiagnosticsForTarget(
       }
     }
 
+
```
nit: seems like a stray change that shouldn't be part of this patch.

```suggestion
```

j-hui commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3637,10 +3636,14 @@ namespace {
-                decl->getLocation());
+            ASTContext &ctx = Impl.SwiftContext;
+            StringRef message =
+                "It should be annotated with either SWIFT_RETURNS_RETAINED or "
+                "SWIFT_RETURNS_UNRETAINED as it is returning a "
+                "SWIFT_SHARED_REFERENCE";
```
nit: wording:

```suggestion
                "This should be annotated with either SWIFT_RETURNS_RETAINED or "
                "SWIFT_RETURNS_UNRETAINED as it returns a SWIFT_SHARED_REFERENCE";
```

j-hui commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3637,10 +3636,14 @@ namespace {
                   Feature::WarnUnannotatedReturnOfCxxFrt)) {
-            Impl.addImportDiagnostic(
-                decl,
-                Diagnostic(diag::no_returns_retained_returns_unretained, decl),
```
Is `no_returns_retained_returns_unretained` still used elsewhere? If not, maybe it should be removed from `DiagnosticsClangImporter.def`.

j-hui reviewed (commented):
I made some nit suggestions, and I'm happy with moving `no_returns_retained_returns_unretained` from a warning to a `@available` annotation. It would be nice if there were a dedicated `@warnOnUse` mechanism so that we're not falsely claiming these APIs are "deprecated", but I'm fine with that being left for future work.

However, I think that given there are two distinct changes involved here (`no_returns_retained_returns_unretained` and `returns_retained_or_returns_unretained_for_non_cxx_frt_values`) , I think they should be made (and reviewed) separately.

> While this diagnostic can be somewhat helpful if a user explicitly annotates a function returning a non-FRT type with SWIFT_RETURNS_(UN)RETAINED, it ends up being inconsistent when we suppress the same diagnostic for templates. In practice, this inconsistency can be confusing — especially since template-based APIs are very common in real-world C++ interop codebases.

I'm not sure I agree here. There are still plenty of C++ APIs that that don't use templates at all, not to mention that these diagnostics should apply to C APIs too if/once we support FRTs from C.

By the way, are there other warnings in `checkBridgingAttrs()`? If so, those should probably be changed too for consistency with `no_returns_retained_returns_unretained`.

egorzhdan reviewed (commented):
Thanks! A couple of minor comments.

fahadnayyar commented:
> I'm not sure I agree here. There are still plenty of C++ APIs that that don't use templates at all, not to mention that these diagnostics should apply to C APIs too if/once we support FRTs from C.

That's an excellent point, John! I now understand why we shouldn't remove the `returns_retained_or_returns_unretained_for_non_cxx_frt_values` error. I was overly focused on C++ templates and neglecting the general applicability. I'll update this patch to to avoid this change.

fahadnayyar commented:
> By the way, are there other warnings in checkBridgingAttrs()? If so, those should probably be changed too for consistency with no_returns_retained_returns_unretained.

We should have a discussion about this! The key difference I observe between `no_returns_retained_returns_unretained` warnings and other diagnostics is that these warnings are intended to guide users towards using annotations for correctness. In contrast, other diagnostics are errors that clearly indicate a mistake in the usage of these annotations, such as using both `SWIFT_RETURNS_RETAINED` and `SWIFT_RETURNS_UNRETAINED` on the same API. Therefore, my feeling is that these errors should be emitted by the clangImporter during the import of the declaration. I believe it would be more effective if these errors were emitted eagerly. What are your thoughts on this?

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

ravikandhadai commented on test/Interop/Cxx/foreign-reference/inheritance-diagnostics.swift:
```diff
@@ -6,34 +6,34 @@
 
 let _ = ImmortalRefereceExample.returnImmortalRefType()
-let _ = ImmortalRefereceExample.returnDerivedFromImmortalRefType()
+let _ = ImmortalRefereceExample.returnDerivedFromImmortalRefType() // expected-warning {{'returnDerivedFromImmortalRefType' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it returns a SWIFT_SHARED_REFERENCE}}
```
I think both of these sites: declaration and use site are interesting. I think we can emit a warning at the declaration site but add note to the use site.

fahadnayyar replied:
I attached note to the call expression and warning to the c++ func decl being called.
@j-hui can you please take a look at the wording of the warning and note message. We want to think about the IDE experience as well. So ideally note should be smaller. But I am not sure how I should phrase the note to be concise as well as convey the information. The current messages actually feel like we are repeting the same things in warning and note.



I’ve attached a note to the call expression and a warning to the C++ function declaration that’s being called. 

@j-hui could you please review the wording of the warning and note message? 
We want to consider the IDE experience as well. Ideally, the note should be smaller. However, I’m not sure how to phrase the note concisely while still conveying the information. The current messages feel like we’re repeating the same things in the warning and note.

```
 warning: 'retrunRefType' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it returns a SWIFT_SHARED_REFERENCE
  60 | template <typename T>
  61 | struct CxxFrtFactory {
  62 | static T* retrunRefType() {return new T;}
     |           `- warning: 'retrunRefType' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it returns a SWIFT_SHARED_REFERENCE
  63 | };
  64 | 

main.swift:63:37: note: Calling C++ function 'retrunRefType' returning a  SWIFT_SHARED_REFERENCE type without annotating it with SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED
  61 | // func foo() {
  62 |     // Templated member method
  63 |     var _ = CxxFrtFactoryOfRefTypeA.retrunRefType()
     |                                     `- note: Calling C++ function 'retrunRefType' returning a  SWIFT_SHARED_REFERENCE type without annotating it with SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED
  64 |     var _ = CxxFrtFactoryOfRefTypeB.retrunRefType()
  65 |     
```

j-hui reviewed (commented):
I'm happy that this is now moved to an AST walker pattern, but I think the implementation of `DiagnoseWalker::walkToExprPre()` is too long and really needs to be broken up into smaller pieces that are easier to reason about.

To me, the obvious logical boundaries/abstractions you could break it into are:
- are we dealing with a Clang-imported function?
- does it return an FRT?
- should it be exempt from this warning?
- emit a diagnostic

I think these can, and probably should, be written as helper functions with concise but descriptive names. I also think setting a Boolean flag like `unannotatedAPIWarningNeeded` makes the logic even harder to interpret, especially since it gets set and unset in different places.

j-hui commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6421,6 +6423,113 @@ static void diagnoseMissingMemberImports(const Expr *E, const DeclContext *DC) {
+              const clang::RecordDecl *recordDecl = recordType->getDecl();
+
+              auto semanticsKind = evaluateOrDefault(
+                  Ctx.evaluator, CxxRecordSemantics({recordDecl, Ctx, nullptr}),
```
It looks like CI is failing because you're not really supposed to pass a `nullptr` here. Then again, the reason that `importerImpl` pointer is needed in the first place is to support emitting diagnostics while performing this query, which I think should be done separately anyway.

fahadnayyar replied:
I’m investigating the CI failure and think about workarounds for now. However, I agree that we should refactor request evaluation to prevent diagnostics from being emitted within the evaluator.

j-hui replied:
I forget exactly what the diagnostic is (it's also related to FRTs), but it's definitely not something we want to see every time we _compute_ the record semantics of a record type. I'd really like that to be removed, but in the meantime I think we'll need to adjust the `CxxRecordSemantics` request to not even attempt to emit the diagnostic when it has a `nullptr`.

j-hui replied:
Oh, it's from here: https://github.com/swiftlang/swift/pull/77522/files#diff-42157eac6817c14ac84af1ca40f706f0b529d927ff3f5a04604e5c7c98d87c80R345

And I might be mistaken about why that was failing. I took a look at the code and it seems like it is correctly checking for `nullptr` before dereferencing.

fahadnayyar replied:
> Is the diagnostic that we get when this is nullptr even important? When do we want to see that diagnostic? I wonder if that is something that we can just remove.

Yes, I did exactly that to resolve the CI issue. The else branch, which was taken when `importerimpl` was `nullptr`, was redundant anyway. 

If the third argument to the `CxxRecordSemanticsDescriptor` constructor is `nullptr`, it suggests that we expect the `evaluateOrDefault` function to avoid emitting a warning when given that descriptor. This pattern is consistent in other instances of calling evaluateOrDefault, such as: `ClassDecl::isForeignReferenceType()`, `MissingMemberFailure::diagnoseUnsafeCxxMethod`, etc.

fahadnayyar commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -192,6 +192,23 @@ bool importer::hasImmortalAttrs(const clang::RecordDecl *decl) {
+  if (decl->hasAttrs()) {
+    for (const auto *attr : decl->getAttrs()) {
+      if (const auto *swiftAttr = llvm::dyn_cast<clang::SwiftAttrAttr>(attr)) {
+        if (swiftAttr->getAttribute() == "returns_unretained") {
```
The `getReturnsUnRetainedAttrInfo` utility function checks for the presence of `returns_(un)retained` `swiftAttr` on any `clang::NamedDecl`. At the call sites of this utility (in `diagnoseCxxFunctionCalls` and `checkBridgingAttrs`), I perform additional checks like `hasImmortalAttrs`. Are you suggesting that we modify this utility to incorporate some of these checks directly into its implementation?

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -192,6 +192,23 @@ bool importer::hasImmortalAttrs(const clang::RecordDecl *decl) {
+importer::ReturnsUnRetainedAttrInfo
+importer::getReturnsUnRetainedAttrInfo(const clang::NamedDecl *decl) {
+  ReturnsUnRetainedAttrInfo info;
+  if (decl->hasAttrs()) {
```
Nit: early return to remove indentation.

fahadnayyar commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6421,6 +6423,115 @@ static void diagnoseMissingMemberImports(const Expr *E, const DeclContext *DC) {
+                    Ctx.LangOpts.hasFeature(
+                        Feature::WarnUnannotatedReturnOfCxxFrt)) {
+
+                  Identifier name;
```
Can you please explain this suggestion in more detail?

If I do:
```
Ctx.Diags.diagnose(
                func->getLoc(), diag::warn_unannotated_cxx_func_returning_frt,
                func);
```
                
instead of  
```
Ctx.Diags.diagnose(func->getLoc(),
                               diag::warn_unannotated_cxx_func_returning_frt,
                               name);
```

I see this compilation error:

```
 error: no matching member function for call to 'diagnose'
 6510 |             Ctx.Diags.diagnose(CE->getLoc(),
      |             ~~~~~~~~~~^~~~~~~~
DiagnosticEngine.h:1021:5: note: candidate function [with ArgTypes = <swift::Identifier>] not viable: no known conversion from 'ValueDecl *' to 'typename detail::PassArgument<Identifier>::type' (aka 'swift::Identifier') for 3rd argument
 1021 |     diagnose(SourceLoc Loc, Diag<ArgTypes...> ID,
      |     ^
 1022 |              typename detail::PassArgument<ArgTypes>::type... Args) {

```

Xazax-hun replied:
You also need to update the diagnostic's definition:
```
WARNING(warn_unannotated_cxx_func_returning_frt, none,
         "%0 should be annotated with either SWIFT_RETURNS_RETAINED or "
         "SWIFT_RETURNS_UNRETAINED as it returns a SWIFT_SHARED_REFERENCE",
         (Identifier))
```

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

Xazax-hun commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6421,6 +6423,112 @@ static void diagnoseMissingMemberImports(const Expr *E, const DeclContext *DC) {
+      if (!func)
+        return Action::Continue(E);
+
+      auto *clangDecl = func->getClangDecl();
+      if (!clangDecl)
+        return Action::Continue(E);
+
+      auto *ND = dyn_cast<clang::NamedDecl>(clangDecl);
+      if (!ND)
+        return Action::Continue(E);
```
```suggestion
      auto *ND = dyn_cast_or_null<clang::NamedDecl>(func->getClangDecl());
      if (!ND)
        return Action::Continue(E);
```

Xazax-hun reviewed (approved):
This looks really clean now, thanks!

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6421,6 +6423,106 @@ static void diagnoseMissingMemberImports(const Expr *E, const DeclContext *DC) {
+    return true;
+  }
+
+  return isa<clang::ObjCMethodDecl>(ND);
```
@j-hui I added some comments explaining these things.

The ownership annotations are not yet supported for constructors (ctor), destructors (dtor), and overloaded operators. These are some enhancements that we may prioritize for the E release.

fahadnayyar commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6421,6 +6423,106 @@ static void diagnoseMissingMemberImports(const Expr *E, const DeclContext *DC) {
+      if (!CE)
+        return Action::Continue(E);
+
+      auto *func = CE->getCalledValue(/*skipFunctionConversions=*/true);
```
The outdated comments still remain on the conversation tab. So I am marking this comment as resolved. I'll reply to the other thread.

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

Xazax-hun reviewed (approved):
I have one nit inline, once that is addressed the PR looks good to me.

fahadnayyar commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6421,6 +6423,133 @@ static void diagnoseMissingMemberImports(const Expr *E, const DeclContext *DC) {
+      if (shouldDiagnoseMissingReturnsRetained(ND, retType, Ctx)) {
+        SourceLoc diagnosticLoc = func->getLoc();
+        if (diagnosticLoc.isInvalid() && func->getClangDecl()) {
+          if (auto *namedDecl =
+                  dyn_cast_or_null<clang::NamedDecl>(func->getClangDecl())) {
```
Removed the `namedDecl` variable, thanks for observing!

@Xazax-hun for `clang::ObjCMethodDecl` I was not able to see correct source location attached to the decl when I tried to emit the warning.

This was the exact example:

```
struct CxxRefType {
} __attribute__((swift_attr("import_reference")))
__attribute__((swift_attr("retain:retainCxxRefType")))
__attribute__((swift_attr("release:releaseCxxRefType")));

struct CxxValType {};

void retainCxxRefType(CxxRefType *_Nonnull b) {}
void releaseCxxRefType(CxxRefType *_Nonnull b) {}

@interface Bridge : NSObject

+ (struct CxxRefType *)objCMethodReturningFRTUnannotated;
@end

@implementation Bridge
+ (struct CxxRefType *)objCMethodReturningFRTUnannotated {
}
@end
```


The warning was not getting emitted correctly at declaration of `objCMethodReturningFRTUnannotated `when I passed `func->getLoc()` to the first argument of `Ctx.Diags.diagnose`

fahadnayyar replied:
@Xazax-hun filed this radar to you: rdar://158977633

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar committed:
[cxx-interop] [Sema] diagnose calls to C++ APIs returning SWIFT_SHARED_REFERENCE without ownership annotations

Move the “unannotated SWIFT_SHARED_REFERENCE return” diagnostic from
ClangImporter to Sema use sites (call expressions). This reduces the
noise from large headers and only warns when the API is actually
used.

- New warning: warn_unannotated_cxx_func_returning_frt
  “cannot infer the ownership of the returned value, annotate %0 with
   either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED”
  plus a note on the declaration site.
- Gated by -enable-experimental-feature WarnUnannotatedReturnOfCxxFrt.

- Implementation:
  * Add diagnoseCxxFunctionCalls() walker in MiscDiagnostics.
  * Add helpers: ReturnOwnershipInfo, hasImmortalAttrs, isReturningFRT().
  * Remove the old import-time GROUPED_WARNING and update tests.

rdar://150800115

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test

fahadnayyar commented:
@swift-ci please smoke test windows platform

fahadnayyar commented:
@swift-ci please smoke test
